pac4j-core/src/main/java/org/pac4j/core/client/IndirectClient2.java

```diff
@@ -0,0 +1,148 @@
+
+    private Redirector redirector;
+
+    private Extractor<C> extractor;
+
+    private Authenticator<C> authenticator;
+
+    private ProfileCreator<C, U> profileCreator =  AuthenticatorProfileCreator.INSTANCE;
+
+    @Override
+    protected void internalInit(final WebContext context) {
+        super.internalInit(context);
+        CommonHelper.assertNotNull("redirector", this.redirector);
+        CommonHelper.assertNotNull("extractor", this.extractor);
+        CommonHelper.assertNotNull("authenticator", this.authenticator);
+        CommonHelper.assertNotNull("profileCreator", this.profileCreator);
+        if (authenticator instanceof InitializableWebObject) {
+            ((InitializableWebObject) this.authenticator).init(context);
+        }
+    }
+
+    @Override
+    protected RedirectAction retrieveRedirectAction(final WebContext context) {
+        return redirector.redirect(context);
+    }
+
+    @Override
+    protected C retrieveCredentials(final WebContext context) throws RequiresHttpAction {
+        try {
+            final C credentials = this.extractor.extract(context);
+            if (credentials == null) {
+                return null;
+            }
+            this.authenticator.validate(credentials);
+            return credentials;
+        } catch (CredentialsException e) {
+            logger.error("Failed to retrieve or validate credentials", e);
+            return null;
+        }
+    }
+
+    @Override
+    protected U retrieveUserProfile(final C credentials, final WebContext context) {
+        final U profile = this.profileCreator.create(credentials);
+        logger.debug("profile: {}", profile);
+        return profile;
+    }
+
+    protected void assertAuthenticatorTypes(final Class<? extends Authenticator>... classes) {
+        if (this.authenticator != null && classes != null) {
+            for (final Class<? extends Authenticator> clazz : classes) {
+                Class<? extends Authenticator> authClazz = this.authenticator.getClass();
+                if (LocalCachingAuthenticator.class.isAssignableFrom(authClazz)) {
+                    authClazz = ((LocalCachingAuthenticator) this.authenticator).getDelegate().getClass();
+                }
+                if (!clazz.isAssignableFrom(authClazz)) {
+                    throw new TechnicalException("Unsupported authenticator type: " + authClazz);
+                }
+            }
+        }
+    }
+
+    @Override
+    public String toString() {
+        return CommonHelper.toString(this.getClass(), "name", getName(), "redirector", this.redirector, "extractor", this.extractor,
+                "authenticator", this.authenticator, "profileCreator", this.profileCreator);
+    }
+
+    public Redirector getRedirector() {
+        return redirector;
+    }
+
+    public void setRedirector(final Redirector redirector) {
+        if (this.redirector == null) {
```

Comment:
Why you would reset your specific one? If you have a specific one, you set
it and it will be used instead of the default one. If you don't have a
specific one, the default one will be used.



2016-02-22 19:10 GMT+01:00 Misagh Moayyed <notifications@github.com>:

> In pac4j-core/src/main/java/org/pac4j/core/client/IndirectClient2.java
> <https://github.com/pac4j/pac4j/pull/497#discussion_r53665874>:
>
> > +            }
> > +        }
> > +    }
> > +
> > +    @Override
> > +    public String toString() {
> > +        return CommonHelper.toString(this.getClass(), "name", getName(), "redirector", this.redirector, "extractor", this.extractor,
> > +                "authenticator", this.authenticator, "profileCreator", this.profileCreator);
> > +    }
> > +
> > +    public Redirector getRedirector() {
> > +        return redirector;
> > +    }
> > +
> > +    public void setRedirector(final Redirector redirector) {
> > +        if (this.redirector == null) {
>
> So if I have set a specific one, I cant reset it? Why? Trying to
> understand the reasoning behind this.
>
> I guess my point is, it may not make sense to set and rest redirectors,
> but leave that decision to the developer.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pac4j/pac4j/pull/497/files#r53665874>.
>